### PR TITLE
Added TCP and UDP support

### LINKED ISSUE #70: feat: support sending serialized valid UDP and TCP packets over IPv4
feat: support sending serialized valid UDP and TCP packets over IPV4

- https://github.com/kytos-ng/sdntrace/blob/master/tracing/trace_pkt.py#L43-L52

- https://github.com/kytos-ng/sdntrace/blob/master/tracing/trace_pkt.py#L160-L195

At the moment neither UDP nor TCP packets are supported, so, when a UDP or TCP payload is passed, the input should be similar to `sdntrace_cp` openapi spec https://github.com/kytos-ng/sdntrace_cp/blob/master/openapi.yml#L59-L95, but then only accept TCP = 6, UDP = 17, and only for IPv4 related fields for now. The data on UDP or TCP can be dummy bytes. src and dst ports should be parametrized too. 

This implementation should work on both OvS and NoviFlow platforms when the PacketOut is sent.

### PR DESCRIPTION
**Pull request by Alopalao**
Closes #70

### Summary

Added TCP and UDP support over a IPv4 header

### Local Tests
Performed some traces.
Additionally set up only TCP/UDP matching flow in a topology.

### End-to-End Tests
From [PR](https://github.com/kytos-ng/kytos-end-to-end-tests/pull/313)
```
+ python3 -m pytest tests/test_e2e_40_sdntrace.py --reruns 2 -r fEr
============================= test session starts ==============================
platform linux -- Python 3.11.2, pytest-8.1.1, pluggy-1.5.0
rootdir: /tests
plugins: rerunfailures-13.0, timeout-2.2.0, anyio-4.3.0
collected 16 items

tests/test_e2e_40_sdntrace.py ................                           [100%]
```

### CODE REVIEW

File: shared/extd_nw_types.py [hunk 1]
```diff
@@ -3,21 +3,257 @@
...
+    sequence_n = UBInt32()
+    acknowledgment_n = UBInt32()
+    _length_flags = UBInt16()
+    window = UBInt16()
+    checksum = UBInt16()
+    urgent_pointer = UBInt16()
+    options = BinaryData()
+    data = BinaryData()
+
+    def __init__(self, source_p:int=0, destination_p:int=0, sequence_n:int=0, acknowledgment_n:int=0, h_length:int=5,
```
Comment (viniarck) [inline shared/extd_nw_types.py]: When you assign the default integer value no need to also type hint with `:int` that 's derived implicitly. Although the intention was good, feel free to remove it.

File: shared/extd_nw_types.py [hunk 2]
```diff
@@ -3,21 +3,257 @@
...
 __all__ = ('TCP', 'UDP')
 
 
 class TCP(GenericStruct):
-    """TCP - To be developed """
-    def __init__(self):
-        pass
+    """TCP - Being developed """
+
+    source_p = UBInt16()
```
Comment (viniarck) [inline shared/extd_nw_types.py]: Let's add this ascii representation in the docstring https://www.ietf.org/rfc/rfc793.txt

```
  TCP Header Format

                                    
    0                   1                   2                   3   
    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |          Source Port          |       Destination Port        |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |                        Sequence Number                        |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |                    Acknowledgment Number                      |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |  Data |           |U|A|P|R|S|F|                               |
   | Offset| Reserved  |R|C|S|S|Y|I|            Window             |
   |       |           |G|K|H|T|N|N|                               |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |           Checksum            |         Urgent Pointer        |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |                    Options                    |    Padding    |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
   |                             data                              |
   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

File: shared/extd_nw_types.py [hunk 3]
```diff
@@ -3,21 +3,257 @@
...
+    destination_p = UBInt16()
```
Comment (viniarck) [inline shared/extd_nw_types.py]: Minor refactoring naming suggestions (these are typically used as abbreviations):

`source_p` -> `src_port`
`destination_p` -> `dst_port`
`sequence_n` -> `seq`
`acknowledgment_n` -> `ack`

File: shared/extd_nw_types.py [hunk 4]
```diff
@@ -3,21 +3,257 @@
...
+        else:
+            self.data = self.options.value
+            self.options = b''
 
 
 class UDP(GenericStruct):
-    """ UDP - To be developed """
+    """ UDP - Being developed """
+
+    source_p = UBInt16()
```
Comment (viniarck) [inline shared/extd_nw_types.py]: Same suggestions `src_port`, and `dst_port`

File: shared/extd_nw_types.py [hunk 5]
```diff
@@ -3,21 +3,257 @@
...
+            self.data = self.options.value[option_size:]
+            self.options = self.options.value[:option_size]
+        else:
+            self.data = self.options.value
+            self.options = b''
 
 
 class UDP(GenericStruct):
-    """ UDP - To be developed """
+    """ UDP - Being developed """
```
Comment (viniarck) [inline shared/extd_nw_types.py]: Let's include the ascii representation https://www.ietf.org/rfc/rfc768.txt in the docstring 


```

                                    
                  0      7 8     15 16    23 24    31  
                 +--------+--------+--------+--------+ 
                 |     Source      |   Destination   | 
                 |      Port       |      Port       | 
                 +--------+--------+--------+--------+ 
                 |                 |                 | 
                 |     Length      |    Checksum     | 
                 +--------+--------+--------+--------+ 
                 |                                     
                 |          data octets ...            
                 +---------------- ...                 

                      User Datagram Header Format
```

**Review by viniarck — CHANGES_REQUESTED** [review]
Serialization-wise looks good, also, map an issue on `python-openflow` to eventually move `TCP` and `UDP` `GenericStruct` there (no need to be done now). 

Locally, it's crashing for me:

```
{
    "trace": {
        "switch": {
            "dpid": "{{dpid}}",
            "in_port": 1
        },
        "eth": {
            "dl_type": 2048
        },
        "ip": {
            "nw_proto": 6
        },
        "tp": {
            "tp_src": 1024,
            "tp_dst": 53
        }
    }
}
```

```
Exception ignored in thread started by: <bound method TraceManager._spawn_trace of <napps.amlight.sdntrace.tracing.trace_manager.TraceManager object at 0x7f41843033d0>>
Traceback (most recent call last):
  File "/home/viniarck/repos/napps/napps/amlight/sdntrace/tracing/trace_manager.py", line 95, in _spawn_trace
    tracer.tracepath()
  File "/home/viniarck/repos/napps/napps/amlight/sdntrace/tracing/tracer.py", line 83, in tracepath
    self.tracepath_loop(entries, color, switch)
  File "/home/viniarck/repos/napps/napps/amlight/sdntrace/tracing/tracer.py", line 120, in tracepath_loop
    entries, color, switch = prepare_next_packet(entries, result,
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/viniarck/repos/napps/napps/amlight/sdntrace/tracing/trace_pkt.py", line 88, in prepare_next_packet
    trace_entries.dl_vlan = _get_vlan_from_pkt(
                            ^^^^^^^^^^^^^^^^^^^
  File "/home/viniarck/repos/napps/napps/amlight/sdntrace/tracing/trace_pkt.py", line 220, in _get_vlan_from_pkt
    vlan = ethernet.vlans[0]
           ~~~~~~~~~~~~~~^^^
IndexError: list index out of range
```

Once e2e is passing with both UDP and TCP flow entries similar to what we spoke, let me know and I'll do a final review. And after that we can book the INT lab to test on NoviFlow switches, but first let's make sure it's all set on OvS. 

Also changelog needs to be updated.

File: shared/extd_nw_types.py (hunk 3 above)
Comment (Alopalao) [reply shared/extd_nw_types.py]: Changed.

File: shared/extd_nw_types.py (hunk 4 above)
Comment (Alopalao) [reply shared/extd_nw_types.py]: Renamed.

File: shared/extd_nw_types.py (hunk 5 above)
Comment (Alopalao) [reply shared/extd_nw_types.py]: Added

File: shared/extd_nw_types.py (hunk 2 above)
Comment (Alopalao) [reply shared/extd_nw_types.py]: Included.

File: shared/extd_nw_types.py (hunk 1 above)
Comment (Alopalao) [reply shared/extd_nw_types.py]: Removed.

**Review by viniarck — APPROVED** [review]
LGTM, excellent to have this feature. Good coverage too. 

I didn't look too much in detail in the packing/unpacking but look coherent to me, and we can keep augmenting or cover more if one day needed, but as of now seems great. Feel free to merge and then map another placeholder task for us to try it out with `telemetry_int`.

**Pull request merged**